### ISSUE TITLE
Resources managed using metacontroller sometimes "hang" when the cache is being refreshed

### ISSUE THREAD

**Issue by maljub01** [metacontroller/metacontroller#78]
First, here's some context on my setup:
* I'm currently using metacontroller on a GKE cluster, which has a per-project rate limit of 10 QPS on the Kubernetes API server.
* Metacontroller on my cluster is managing a lot of resources (~1000 parent resources).
* Although ultimately, GKE's rate limit is the primary cause behind the issues I'm seeing, I suspect this is still something that metacontroller might want to handle since they probably can still be triggered on a cluster with a higher rate limit by proportionally increasing the number of resources.

This leads to what looks like resources failing to update whenever metacontroller is syncing that resource's cache.

Here's an example of how this happens:
* A parent resource has a running pod child resource.
* Metacontroller starts syncing its caches (either because it was redeployed or because the caches expired). This results in the parent controller's work queue immediately becoming very large (all 1000+ resources are enqueued).
* GKE throttles the requests from metacontroller, which means that the parent controller takes a long time to work through its  queue.
* When the pod child resource finishes running, this gets detected by metacontroller and the parent resource gets enqueued.
* However, it takes a long time before metacontroller gets around to actually syncing the resource (I've had cases where it took 20+ minutes). This makes it look like the parent resources or the metacontroller are stuck in a bad state.

Given this issue, I am wondering:
* Why does metacontroller clear its cache periodically? Is this safe to disable? I'm currently mitigating this issue is by setting an unrealistically large cache flush interval (~100 years), but ideally I'd like to simply disable it as long as that doesn't break things.
* If periodically clearing the cache is necessary or useful, is it possible to instead do it on a per-item basis (ie. a value that's 30 minutes old gets cleared and refreshed rather than all cache values at every 30 minute interval)?
* Similarly, why does flushing the cache also result in immediately rebuilding the entire cache and syncing everything? Why not simply clear the cache and let it get rebuilt as things continue to get processed?
* Is it possible to prioritize syncing based on a change to a resource's child over syncing based on changes to the cache? This  change might be as simple as changing the event handler functions to insert syncs from the event handler at the front of the queue rather than adding it at the back of the queue.

**Comment by fpetkovski** [comment]
Hey, I can provide some of my own experience here in the hope it would help you:

- metacontroller relies on the informer component which comes as part of the kubernetes client. This component is designed to flush the cache periodically and relist objects from the API server to avoid a long-lived stale cache. It is defined [here](https://github.com/metacontroller/metacontroller/blob/master/main.go#L43)
- I don't think this is easily doable due to #1
- flushing the cache means relisting everything from the API server. I assume this also re-triggers the reconciliation loop for every item in the cache. This is how kubernetes controllers avoid drift between desired and observed state.
- I think this should be doable, I assume that strategy should be implemented [here](https://github.com/metacontroller/metacontroller/blob/master/controller/composite/controller.go#L292)

However I would also point out that your webhook might not be fully idempotent. Metacontroller can detect when a sync will not change anything in Kubernetes and will not make any API requests as you can see [here](https://github.com/metacontroller/metacontroller/blob/master/controller/common/manage_children.go#L205). Hope this helps.

**Comment by maljub01** [comment]
Thanks @fpetkovski for looking into this and for your responses. Sorry it took me some time to get back to you. I've been investigating this issue more closely and trying to figure out how to fix it.

Here's what I found out:
* It turned out the issue I was seeing was not caused by the Kubernetes API throttling requests, rather it was caused by metacontroller having only 5 workers for processing the queue, which is specified [here](https://github.com/metacontroller/metacontroller/blob/b686a7f03fbc301e3a87c0e1c35cbb3630a1445f/controller/composite/controller.go#L181).
* I modified metacontroller to make it possible to specify the number of workers as a command line flag, once I increased that (I used 1000), metacontroller went from taking 40-50 minutes every time the cache is flushed to being able to resolve everything within seconds.

I'll send a PR with the change that adds a flag for controlling the number of workers, but I also suspect it might be a good idea to consider adding a similar check to the one you mentioned in your comment where metacontroller avoids making sync requests for objects that have not changed since the last one.
